pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go

```diff
@@ -135,33 +141,49 @@ func (og *operationGenerator) GenerateRegisterPluginFunc(
 func (og *operationGenerator) GenerateUnregisterPluginFunc(
 	socketPath string,
 	pluginHandlers map[string]cache.PluginHandler,
+	pathToHandlers map[string]cache.HandlerWithName,
 	actualStateOfWorldUpdater ActualStateOfWorldUpdater) func() error {
 
 	unregisterPluginFunc := func() error {
-		client, conn, err := dial(socketPath, dialTimeoutDuration)
-		if err != nil {
-			return fmt.Errorf("UnregisterPlugin error -- dial failed at socket %s, err: %v", socketPath, err)
+		var pair cache.HandlerWithName
+		var handler cache.PluginHandler
+		ok := false
+		if len(pathToHandlers) != 0 {
+			pair, ok = pathToHandlers[socketPath]
 		}
-		defer conn.Close()
+		if !ok {
+			client, conn, err := dial(socketPath, dialTimeoutDuration)
+			if err != nil {
+				return fmt.Errorf("UnregisterPlugin error -- dial failed at socket %s, err: %v", socketPath, err)
+			}
+			defer conn.Close()
 
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-		defer cancel()
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
 
-		infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
-		if err != nil {
-			return fmt.Errorf("UnregisterPlugin error -- failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
-		}
+			infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
+			if err != nil {
+				return fmt.Errorf("UnregisterPlugin error -- failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+			}
 
-		handler, ok := pluginHandlers[infoResp.Type]
-		if !ok {
-			return fmt.Errorf("UnregisterPlugin error -- no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath)
-		}
+			handler, ok = pluginHandlers[infoResp.Type]
+			if !ok {
+				return fmt.Errorf("UnregisterPlugin error -- no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath)
+			}
+
+			// We remove the plugin to the actual state of world cache before calling a plugin consumer's Unregister handle
+			// so that if we receive a register event during Register Plugin, we can process it as a Register call.
+			actualStateOfWorldUpdater.RemovePlugin(socketPath)
 
-		// We remove the plugin to the actual state of world cache before calling a plugin consumer's Unregister handle
-		// so that if we receive a register event during Register Plugin, we can process it as a Register call.
-		actualStateOfWorldUpdater.RemovePlugin(socketPath)
+			handler.DeRegisterPlugin(infoResp.Name)
+		} else {
+			// We remove the plugin to the actual state of world cache before calling a plugin consumer's Unregister handle
+			// so that if we receive a register event during Register Plugin, we can process it as a Register call.
+			actualStateOfWorldUpdater.RemovePlugin(socketPath)
```

Comment:
can refactor this code to reduce duplication since its at the end of both `if` and `else` blocks. You can just drop the `else`

Reply:
The if block is no longer needed since socket path would be absent.

Reply:
looks like the logic changed entirely. Did we need to do that original `RemovePlugin` if `!handlerFound` or not?

Reply:
In normal cases, handlerFound would be true.
The "unable to DeRegister" shouldn't happen.

I think RemovePlugin() should be paired with DeRegisterPlugin() call - otherwise the internal state would be inconsistent.